### DISCUSSION TITLE
Narrowing types using middleware

### DISCUSSION THREAD

**Discussion by Roardom** [larastan/larastan#1903] [Q&A]
Hello!

I've managed to successfully increase the phpstan level of a large open source project to level 7 so far which has been a big help (thank you!), but I've been having some difficulty getting it to level 8 due to the inconvenience (read: ocd) of adding `abort_if($request->user() === null, 500);` throughout hundreds of routes. I will often either dependency inject form requests and use `$request->user()`, or use `auth()->user()` which both return a `?User` type. But, even though I use the default `Illuminate\Auth\Middleware\Authenticate` middleware to throw exceptions if it's ever null, from what I understand, the middleware isn't used to narrow the type down to `User`, causing 1000+ errors on level 8 compared to level 7.

Is such a task possible with larastan or phpstan? I've never worked directly on phpstan or larastan before, but I would be open to spending a weekend or 2 on this depending on how difficult such a task might be. Is there a recommended place to start?

Thanks!

**Comment by calebdw** [comment]
Middleware can't be used to narrow types because Larastan does not execute the request lifecycle.

One thing you could do is extend the Request and override the `user` method (or create a `userNotNull` method):

```php
class Request extends IlluminateRequest
{
    public function user(): User
    {
        $user = parent::user();

        abort_if($user === null, 500);
      
        return $user;
    }

    // or
    public function userNotNull(): User
    {
        $user = $this->user();

        abort_if($user === null, 500);
      
        return $user;
    }
}
```

**Reply by Roardom** [reply]
Thank you for the feedback!

This is a great idea and is good for silencing the false positives, however I've had issues in the past where I've unintentionally used `$request->user()` inside one of the few controller actions that didn't have the authentication middleware which has bit me quite painfully and I'm trying my best to avoid that happening again.

Based on your information about the lifecycle not being ran, I spent the past few hours reading phpstan docs and came up with the following: https://github.com/Roardom/larastan/commit/d7874a852196fe18bb4deeecc71253fae73b7018

It's not generic in the way that it narrows middleware, it only works with the auth models for the time being. It's not a good enough solution to PR yet, since I'm manually guessing the current route using our codebase conventions but not everyone uses our conventions so that won't work for everyone (`$router->getCurrentRoute()` doesn't work).

Based off this, do you have any ideas on increasing reliability?

**Reply by calebdw** [reply]
Very interesting! However, this will only work for code that is inside a controller---if the controller calls another class that then resolves the auth user then that won't get picked up.

Also, that concern is used for both the Auth::user() and the Request::user() methods so the other one should be updated as well, and of course tests would be needed for an actual PR.

**Reply by Roardom** [reply]
I've been going through the 1300+ errors to upgrade from level 7 to level 8 with the help of this code (it reduces error count to 800, and the model `->withDefault()` fix reduces it to ~600 real errors), but discovered an issue with it. Found an edge case where one of the controller actions is called by both an authenticated route, as well as a route that relies on an apikey without an auth guard. Currently it marks the action as authenticated since at least one route linking to the action is authenticated, but it will need to be reworked so that it checks all routes. Figured I'd note it down here.

**Comment by szepeviktor** [comment]
You could add a stubs file that "fixes" the return type.

**Reply by Roardom** [reply]
Thank you for this idea! I'm actually hoping for a "proper" solution to prevent other errors from happening in the future. Do you mind taking a look at the thread in https://github.com/larastan/larastan/discussions/1903#discussioncomment-9028460 and how the code could be improved?
